v1/topdown/strings.go

```diff
@@ -153,33 +155,48 @@ func builtinConcat(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term)
 		return err
 	}
 
-	strs := []string{}
+	var strs []string
 
 	switch b := operands[1].Value.(type) {
 	case *ast.Array:
-		err := b.Iter(func(x *ast.Term) error {
-			s, ok := x.Value.(ast.String)
+		var l int
+		for i := 0; i < b.Len(); i++ {
+			s, ok := b.Elem(i).Value.(ast.String)
 			if !ok {
-				return builtins.NewOperandElementErr(2, operands[1].Value, x.Value, "string")
+				return builtins.NewOperandElementErr(2, operands[1].Value, b.Elem(i).Value, "string")
 			}
-			strs = append(strs, string(s))
-			return nil
-		})
-		if err != nil {
-			return err
+			l += len(string(s))
+		}
+
+		if b.Len() == 1 {
+			return iter(b.Elem(0))
 		}
+
+		strs = make([]string, 0, l)
+		for i := 0; i < b.Len(); i++ {
+			strs = append(strs, string(b.Elem(i).Value.(ast.String)))
+		}
+
 	case ast.Set:
-		err := b.Iter(func(x *ast.Term) error {
-			s, ok := x.Value.(ast.String)
+		var l int
+		terms := b.Slice()
+		for i := 0; i < len(terms); i++ {
+			s, ok := terms[i].Value.(ast.String)
 			if !ok {
-				return builtins.NewOperandElementErr(2, operands[1].Value, x.Value, "string")
+				return builtins.NewOperandElementErr(2, operands[1].Value, terms[i].Value, "string")
 			}
-			strs = append(strs, string(s))
-			return nil
-		})
-		if err != nil {
-			return err
+			l += len(string(s))
+		}
+
+		if b.Len() == 1 {
+			return iter(b.Slice()[0])
+		}
+
+		strs = make([]string, 0, l)
+		for i := 0; i < b.Len(); i++ {
+			strs = append(strs, string(terms[i].Value.(ast.String)))
 		}
+
 	default:
 		return builtins.NewOperandTypeErr(2, operands[1].Value, "set", "array")
 	}
```

Comment:
Hm.. Since we call strings.Join below anyways, which uses strings.Builder{}, I suppose we could go with strings.Builder right away. But it probably doesn't make much of a difference, and this is cleaner.

Reply:
Not a bad idea. I might try that later, but will leave as is for now.